{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6fbd0810-1c3d-4b2f-8c76-afc036bde78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import sys\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "4303dd1b-3ba9-4e92-a075-808636121873",
   "metadata": {},
   "outputs": [],
   "source": [
    "accession=\"GCA_000001405.29\"\n",
    "threshold_short=1000000\n",
    "threshold_ends=0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0ce7320d-3916-4cda-aca1-1788566c917d",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_directory = f\"/lab/wengpj01/vertebrate_pipeline/subdirs/{accession}\"\n",
    "os.chdir(new_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "d760d509-0fd2-40eb-8ac1-7103f84310fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read chromsizes.csv into a pandas dataframe\n",
    "chromsizes = pd.read_csv('chromsizes.csv', sep='\\t', header=None, names=['chr', 'length'])\n",
    "\n",
    "# Find and read all *_from_pipeline.gtf files into a list of dataframes\n",
    "gtf_files = glob.glob('*_from_pipeline.gtf')\n",
    "gene_dfs = []\n",
    "\n",
    "for gtf_file in gtf_files:\n",
    "    df = pd.read_csv(gtf_file, sep='\\t', header=None, names=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'])\n",
    "    gene_dfs.append(df)\n",
    "    \n",
    "    # Strip text before the last underscore in the 'seqname' column\n",
    "    df['seqname'] = df['seqname'].str.replace(r'^.*_', '', regex=True)\n",
    "\n",
    "# Concatenate all dataframes into a single dataframe\n",
    "genes = pd.concat(gene_dfs, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "52d41b65-dfb3-4b5b-925b-da8edfc84602",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through rows of the 'genes' dataframe and find corresponding chromosome length\n",
    "genes['chr_length'] = None  # Create an empty column to store chromosome lengths\n",
    "genes['chr_mini'] = None\n",
    "genes['ends'] = None\n",
    "\n",
    "for i, gene_row in genes.iterrows():\n",
    "    seqname = gene_row['seqname']\n",
    "    \n",
    "    # Find the corresponding row in 'chromsizes'\n",
    "    chrom_row = chromsizes[chromsizes['chr'] == seqname]\n",
    "    \n",
    "    if not chrom_row.empty:\n",
    "        chr_length = chrom_row.iloc[0]['length']\n",
    "        genes.at[i, 'chr_length'] = chr_length\n",
    "        \n",
    "        to_end = (chr_length - gene_row[\"start\"])/chr_length\n",
    "        to_start = (gene_row[\"start\"])/chr_length\n",
    "        \n",
    "        if (to_start < threshold_ends) or (to_end < threshold_ends):\n",
    "            genes.at[i, 'ends'] = \"True\"\n",
    "        else:\n",
    "            genes.at[i, 'ends'] = \"False\"\n",
    "        \n",
    "        if chr_length < threshold_short:\n",
    "            genes.at[i, 'chr_mini'] = \"True\"\n",
    "        else:\n",
    "            genes.at[i, 'chr_mini'] = \"False\"\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "643d434e-454f-4e1d-a4f7-b8a6a70c9c76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5333333333333333\n"
     ]
    }
   ],
   "source": [
    "# Calculate the fraction of rows where 'chr_mini' is True\n",
    "fraction_chr_mini_true = len(genes[genes['chr_mini'] == \"True\"]) / len(genes)\n",
    "print(fraction_chr_mini_true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "74c628ca-5e7a-4677-a0d7-d2eb08de05f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.3333333333333333\n"
     ]
    }
   ],
   "source": [
    "# Calculate the fraction of rows where 'ends' is True\n",
    "fraction_near_ends = len(genes[genes['ends'] == \"True\"]) / len(genes)\n",
    "print(fraction_near_ends)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "89f98877-25ff-47e2-87d4-ccbed3090b09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5\n"
     ]
    }
   ],
   "source": [
    "# Calculate the fraction of rows where 'ends' is True and 'chr_mini' is False\n",
    "fraction_near_ends_long = len(genes[(genes['ends'] == \"True\") & (genes['chr_mini'] == \"False\")]) / len(genes[genes['chr_mini'] == \"False\"])\n",
    "print(fraction_near_ends_long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "b39a92b5-5876-4532-a9a3-8a12509ee40d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>chr_length</th>\n",
       "      <th>chr_mini</th>\n",
       "      <th>ends</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CM000667.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>9629133</td>\n",
       "      <td>9630032</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH9511\"; transcript_id \"KH9511.1\"</td>\n",
       "      <td>181538259</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>122994759</td>\n",
       "      <td>122995634</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH8006\"; transcript_id \"KH8006.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>141764159</td>\n",
       "      <td>141765109</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH3992\"; transcript_id \"KH3992.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>141778489</td>\n",
       "      <td>141779388</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH3483\"; transcript_id \"KH3483.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>141790101</td>\n",
       "      <td>141791261</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH5175\"; transcript_id \"KH5175.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>141972688</td>\n",
       "      <td>141973689</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH3818\"; transcript_id \"KH3818.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>143183419</td>\n",
       "      <td>143184435</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH11253\"; transcript_id \"KH11253.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CM000669.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>143222079</td>\n",
       "      <td>143223050</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH2887\"; transcript_id \"KH2887.1\"</td>\n",
       "      <td>159345973</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>10806051</td>\n",
       "      <td>10806980</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH1732\"; transcript_id \"KH1732.1\"</td>\n",
       "      <td>133275309</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>10938254</td>\n",
       "      <td>10939222</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH11772\"; transcript_id \"KH11772.1\"</td>\n",
       "      <td>133275309</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>10996946</td>\n",
       "      <td>10997875</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH1646\"; transcript_id \"KH1646.1\"</td>\n",
       "      <td>133275309</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>11061365</td>\n",
       "      <td>11062294</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH5043\"; transcript_id \"KH5043.1\"</td>\n",
       "      <td>133275309</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>11091300</td>\n",
       "      <td>11092478</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH8136\"; transcript_id \"KH8136.1\"</td>\n",
       "      <td>133275309</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>CM000674.2</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>11185993</td>\n",
       "      <td>11186937</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH8534\"; transcript_id \"KH8534.1\"</td>\n",
       "      <td>133275309</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GL877876.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>107093</td>\n",
       "      <td>108004</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH7083\"; transcript_id \"KH7083.1\"</td>\n",
       "      <td>408271</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GL877876.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>220378</td>\n",
       "      <td>221334</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH10927\"; transcript_id \"KH10927.1\"</td>\n",
       "      <td>408271</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GL877876.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>229112</td>\n",
       "      <td>230041</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH5582\"; transcript_id \"KH5582.1\"</td>\n",
       "      <td>408271</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GL877876.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>290027</td>\n",
       "      <td>291205</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH8438\"; transcript_id \"KH8438.1\"</td>\n",
       "      <td>408271</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GL877876.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>320</td>\n",
       "      <td>1276</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH901\"; transcript_id \"KH901.1\"</td>\n",
       "      <td>408271</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GL877876.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>363939</td>\n",
       "      <td>364898</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH6198\"; transcript_id \"KH6198.1\"</td>\n",
       "      <td>408271</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>184654</td>\n",
       "      <td>185553</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH7855\"; transcript_id \"KH7855.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>195641</td>\n",
       "      <td>196570</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH1\"; transcript_id \"KH1.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>220366</td>\n",
       "      <td>221322</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH10727\"; transcript_id \"KH10727.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>229411</td>\n",
       "      <td>230340</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH7248\"; transcript_id \"KH7248.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>24052</td>\n",
       "      <td>24975</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH4204\"; transcript_id \"KH4204.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>291887</td>\n",
       "      <td>292786</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH348\"; transcript_id \"KH348.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>328319</td>\n",
       "      <td>329278</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH6402\"; transcript_id \"KH6402.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>KI270904.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>7843</td>\n",
       "      <td>8781</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH9144\"; transcript_id \"KH9144.1\"</td>\n",
       "      <td>572349</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>KZ208912.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>62817</td>\n",
       "      <td>63773</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH3324\"; transcript_id \"KH3324.1\"</td>\n",
       "      <td>589656</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>KZ208912.1</td>\n",
       "      <td>Genome</td>\n",
       "      <td>exon</td>\n",
       "      <td>97237</td>\n",
       "      <td>98160</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"KH2737\"; transcript_id \"KH2737.1\"</td>\n",
       "      <td>589656</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname  source feature      start        end score strand  frame  \\\n",
       "0   CM000667.2  Genome    exon    9629133    9630032     .      -      0   \n",
       "1   CM000669.2  Genome    exon  122994759  122995634     .      -      0   \n",
       "2   CM000669.2  Genome    exon  141764159  141765109     .      +      0   \n",
       "3   CM000669.2  Genome    exon  141778489  141779388     .      +      0   \n",
       "4   CM000669.2  Genome    exon  141790101  141791261     .      +      0   \n",
       "5   CM000669.2  Genome    exon  141972688  141973689     .      -      0   \n",
       "6   CM000669.2  Genome    exon  143183419  143184435     .      +      0   \n",
       "7   CM000669.2  Genome    exon  143222079  143223050     .      +      0   \n",
       "8   CM000674.2  Genome    exon   10806051   10806980     .      -      0   \n",
       "9   CM000674.2  Genome    exon   10938254   10939222     .      -      0   \n",
       "10  CM000674.2  Genome    exon   10996946   10997875     .      -      0   \n",
       "11  CM000674.2  Genome    exon   11061365   11062294     .      -      0   \n",
       "12  CM000674.2  Genome    exon   11091300   11092478     .      -      0   \n",
       "13  CM000674.2  Genome    exon   11185993   11186937     .      -      0   \n",
       "14  GL877876.1  Genome    exon     107093     108004     .      -      0   \n",
       "15  GL877876.1  Genome    exon     220378     221334     .      -      0   \n",
       "16  GL877876.1  Genome    exon     229112     230041     .      -      0   \n",
       "17  GL877876.1  Genome    exon     290027     291205     .      -      0   \n",
       "18  GL877876.1  Genome    exon        320       1276     .      -      0   \n",
       "19  GL877876.1  Genome    exon     363939     364898     .      -      0   \n",
       "20  KI270904.1  Genome    exon     184654     185553     .      -      0   \n",
       "21  KI270904.1  Genome    exon     195641     196570     .      -      0   \n",
       "22  KI270904.1  Genome    exon     220366     221322     .      -      0   \n",
       "23  KI270904.1  Genome    exon     229411     230340     .      -      0   \n",
       "24  KI270904.1  Genome    exon      24052      24975     .      -      0   \n",
       "25  KI270904.1  Genome    exon     291887     292786     .      -      0   \n",
       "26  KI270904.1  Genome    exon     328319     329278     .      -      0   \n",
       "27  KI270904.1  Genome    exon       7843       8781     .      -      0   \n",
       "28  KZ208912.1  Genome    exon      62817      63773     .      +      0   \n",
       "29  KZ208912.1  Genome    exon      97237      98160     .      +      0   \n",
       "\n",
       "                                       attribute chr_length chr_mini   ends  \n",
       "0     gene_id \"KH9511\"; transcript_id \"KH9511.1\"  181538259    False   True  \n",
       "1     gene_id \"KH8006\"; transcript_id \"KH8006.1\"  159345973    False  False  \n",
       "2     gene_id \"KH3992\"; transcript_id \"KH3992.1\"  159345973    False  False  \n",
       "3     gene_id \"KH3483\"; transcript_id \"KH3483.1\"  159345973    False  False  \n",
       "4     gene_id \"KH5175\"; transcript_id \"KH5175.1\"  159345973    False  False  \n",
       "5     gene_id \"KH3818\"; transcript_id \"KH3818.1\"  159345973    False  False  \n",
       "6   gene_id \"KH11253\"; transcript_id \"KH11253.1\"  159345973    False  False  \n",
       "7     gene_id \"KH2887\"; transcript_id \"KH2887.1\"  159345973    False  False  \n",
       "8     gene_id \"KH1732\"; transcript_id \"KH1732.1\"  133275309    False   True  \n",
       "9   gene_id \"KH11772\"; transcript_id \"KH11772.1\"  133275309    False   True  \n",
       "10    gene_id \"KH1646\"; transcript_id \"KH1646.1\"  133275309    False   True  \n",
       "11    gene_id \"KH5043\"; transcript_id \"KH5043.1\"  133275309    False   True  \n",
       "12    gene_id \"KH8136\"; transcript_id \"KH8136.1\"  133275309    False   True  \n",
       "13    gene_id \"KH8534\"; transcript_id \"KH8534.1\"  133275309    False   True  \n",
       "14    gene_id \"KH7083\"; transcript_id \"KH7083.1\"     408271     True  False  \n",
       "15  gene_id \"KH10927\"; transcript_id \"KH10927.1\"     408271     True  False  \n",
       "16    gene_id \"KH5582\"; transcript_id \"KH5582.1\"     408271     True  False  \n",
       "17    gene_id \"KH8438\"; transcript_id \"KH8438.1\"     408271     True  False  \n",
       "18      gene_id \"KH901\"; transcript_id \"KH901.1\"     408271     True   True  \n",
       "19    gene_id \"KH6198\"; transcript_id \"KH6198.1\"     408271     True  False  \n",
       "20    gene_id \"KH7855\"; transcript_id \"KH7855.1\"     572349     True  False  \n",
       "21          gene_id \"KH1\"; transcript_id \"KH1.1\"     572349     True  False  \n",
       "22  gene_id \"KH10727\"; transcript_id \"KH10727.1\"     572349     True  False  \n",
       "23    gene_id \"KH7248\"; transcript_id \"KH7248.1\"     572349     True  False  \n",
       "24    gene_id \"KH4204\"; transcript_id \"KH4204.1\"     572349     True   True  \n",
       "25      gene_id \"KH348\"; transcript_id \"KH348.1\"     572349     True  False  \n",
       "26    gene_id \"KH6402\"; transcript_id \"KH6402.1\"     572349     True  False  \n",
       "27    gene_id \"KH9144\"; transcript_id \"KH9144.1\"     572349     True   True  \n",
       "28    gene_id \"KH3324\"; transcript_id \"KH3324.1\"     589656     True  False  \n",
       "29    gene_id \"KH2737\"; transcript_id \"KH2737.1\"     589656     True  False  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7602c20a-3f41-459b-9444-c91e6364601d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e191b44-caa4-4507-8d80-26c0f90666e5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
